README.md
# Is Your Node Process Too Busy?

[![Build Status](https://secure.travis-ci.org/lloyd/node-toobusy.png)](http://travis-ci.org/lloyd/node-toobusy)

What happens when your service is overwhelmed with traffic?
Your server can do one of two things:

  * Stop working, or...
  * Keep serving as many requests as possible

This library helps you do the latter.

## How it works

`toobusy` polls the node.js event loop and keeps track of "lag",
which is long requests wait in node's event queue to be processed.
When lag crosses a threshold, `toobusy` tells you that you're *too busy*.
At this point you can stop request processing early
(before you spend too much time on them and compound the problem),
and return a "Server Too Busy" response.
This allows your server to stay *responsive* under extreme load,
and continue serving as many requests as possible.

## installation

```
npm install toobusy
```


## usage

```javascript
var toobusy = require('toobusy'),
    express = require('express');

var app = express();

// middleware which blocks requests when we're too busy
app.use(function(req, res, next) {
  if (toobusy()) {
    res.send(503, "I'm busy right now, sorry.");
  } else {
    next();
  }
});

app.get('/', function(req, res) {
  // processing the request requires some work!
  var i = 0;
  while (i < 1e5) i++;
  res.send("I counted to " + i);
});

var server = app.listen(3000);

process.on('SIGINT', function() {
  server.close();
  // calling .shutdown allows your process to exit normally
  toobusy.shutdown();
  process.exit();
});
```

## tuneable parameters

The one knob that the library exposes is "maximum lag".
This number represents the maximum amount of time in milliseconds that the event queue is behind,
before we consider the process *too busy*.

```javascript
// set maximum lag to an aggressive value
require('toobusy').maxLag(10);
```

The default value is 70ms,
which allows an "average" server to run at 90-100% CPU
and keeps request latency at around 200ms.
For comparison, a value of 10ms results in 60-70% CPU usage,
while latency for "average" requests stays at about 40ms.

These numbers are only examples,
and the specifics of your hardware and application can change them drastically,
so experiment!
The default of 70 should get you started.

## references

> There is nothing new under the sun. (Ecclesiastes 1:9)

Though applying "event loop latency" to node.js was not directly inspired by anyone else's work,
this concept is not new.  Here are references to others who apply the same technique:

  * [Provos, Lever, and Tweedie 2000](http://www.kegel.com/c10k.html#tips) - "notes that dropping incoming connections when the server is overloaded improved the shape of the performance curve."

## license

[WTFPL](http://www.wtfpl.net/)


ChangeLog
0.2.4 -
  * fix repository key in package.json

0.2.3 -
  * node.js 0.10.0 support - upgrade to latest bindings and include in .travis.yml

0.2.2 -
  * *really* works on windows
  * improved documentation

0.2.1 -
  * works on windows
  * improved express.js example (now targets express 3.0)

0.2.0 -
  * documentation improvements
  * improved examples (including load generation tool and server)
  * .lag() added to retrieve (periodically calculated) event loop lag in ms

0.1.1 -
  * hey look, unit tests!

0.1.0 -
  * improved documentation and samples
  * added `maxLag` parameter


tests.js
const
should = require('should'),
toobusy = require('./');

describe('the library', function() {
  it('should export a couple functions', function(done) {
    (toobusy).should.be.a.Function;
    (toobusy.maxLag).should.be.a.Function;
    (toobusy.shutdown).should.be.a.Function;
    done();
  });
});

describe('maxLag', function() {
  it('should default to 70', function(done) {
    (toobusy.maxLag()).should.equal(70);
    done();
  });
  it('should throw an exception for values < 10', function(done) {
    (function() { toobusy.maxLag(9); }).should.throw;
    done();
  });
  it('should be configurable', function(done) {
    (toobusy.maxLag(50)).should.equal(50);
    (toobusy.maxLag(10)).should.equal(10);
    (toobusy.maxLag()).should.equal(10);
    done();
  });
});

describe('toobusy()', function() {
  it('should return true after a little load', function(done) {
    function load() {
      if (toobusy()) return done();
      var start = new Date();
      while ((new Date() - start) < 300) {
        for (var i = 0; i < 1e5;) i++;
      }
      setTimeout(load, 0);
    }
    load();
  });

  it('should return a lag value after a little load', function(done) {
    function load() {
      if (toobusy()) {
        var lag = toobusy.lag();
        should.exist(lag);
        lag.should.be.above(1);
        return done();
      }
      var start = new Date();
      while ((new Date() - start) < 250) {
        for (var i = 0; i < 1e5;) i++;
      }
      setTimeout(load, 0);
    }
    load();
  });
});


toobusy.cc
#include <node.h>
#include <nan.h>
#include <uv.h>
#include <stdlib.h>
#if defined(_WIN32)
  #include <time.h>
#else
  #include <sys/time.h>
#endif

using namespace v8;

static const unsigned int POLL_PERIOD_MS = 500;
static unsigned int HIGH_WATER_MARK_MS = 70;
// A dampening factor.  When determining average calls per second or
// current lag, we weigh the current value against the previous value 2:1
// to smooth spikes.
static const unsigned int AVG_DECAY_FACTOR = 3;

//static uv_idle_t s_idler;
static uv_timer_t s_timer;
static uint32_t s_currentLag;
static uint64_t s_lastMark;

NAN_METHOD(TooBusy) {
    NanScope();

    bool block = false;
    if (s_currentLag > HIGH_WATER_MARK_MS) {
        // probabilistically block requests proportional to how
        // far behind we are.
        double pctToBlock = ((s_currentLag - HIGH_WATER_MARK_MS) /
                             (double) HIGH_WATER_MARK_MS) * 100.0;
        double r = (rand() / (double) RAND_MAX) * 100.0;
        if (r < pctToBlock) block = true;
    }
    NanReturnValue(block ? NanTrue() : NanFalse());
}

NAN_METHOD(ShutDown)  {
    // No HandleScope required, because this function allocates no
    // v8 classes that reside on the heap.

    uv_timer_stop(&s_timer);
    NanReturnUndefined();
}

NAN_METHOD(Lag) {
    NanScope();
    NanReturnValue(NanNew<Integer>(s_currentLag));
}

NAN_METHOD(HighWaterMark) {
    NanScope();

    if (args.Length() >= 1) {
        if (!args[0]->IsNumber()) {
            NanThrowError("expected numeric first argument");
            NanReturnUndefined();
        }
        int hwm = args[0]->Int32Value();
        if (hwm < 10) {
            NanThrowError("maximum lag should be greater than 10ms");
            NanReturnUndefined();
        }
        HIGH_WATER_MARK_MS = hwm;
    }

    NanReturnValue(NanNew<Number>(HIGH_WATER_MARK_MS));
}

static void every_second(uv_timer_t* handle)
{
    uint64_t now = uv_hrtime();

    if (s_lastMark > 0) {
        // keep track of (dampened) average lag.
        uint32_t lag = (uint32_t) ((now - s_lastMark) / 1000000);
        lag = (lag < POLL_PERIOD_MS) ? 0 : lag - POLL_PERIOD_MS;
        s_currentLag = (lag + (s_currentLag * (AVG_DECAY_FACTOR-1))) /
            AVG_DECAY_FACTOR;
    }
    s_lastMark = now;
}

static void every_second(uv_timer_t* handle, int status)
{
    every_second(handle);
}

extern "C" void InitAll(Handle<Object> exports) {
    exports->Set(NanNew<String>("toobusy"),
      NanNew<FunctionTemplate>(TooBusy)->GetFunction());
    exports->Set(NanNew<String>("shutdown"),
      NanNew<FunctionTemplate>(ShutDown)->GetFunction());
    exports->Set(NanNew<String>("lag"),
        NanNew<FunctionTemplate>(Lag)->GetFunction());
    exports->Set(NanNew<String>("maxLag"),
      NanNew<FunctionTemplate>(HighWaterMark)->GetFunction());

    uv_timer_init(uv_default_loop(), &s_timer);
    uv_timer_start(&s_timer, every_second, POLL_PERIOD_MS, POLL_PERIOD_MS);
};

NODE_MODULE(toobusy, InitAll);


toobusy.js
var STANDARD_HIGHWATER = 70;
var STANDARD_INTERVAL = 500;

// A dampening factor.  When determining average calls per second or
// current lag, we weigh the current value against the previous value 2:1
// to smooth spikes.
var AVG_DECAY_FACTOR = 3;

var lastTime = new Date().valueOf(), now, lag, highWater = STANDARD_HIGHWATER, interval = STANDARD_INTERVAL, currentLag = 0;

var checkInterval = setInterval(function(){
  now = new Date().valueOf();
  lag = now - lastTime;
  lag = (lag < interval) ? 0 : lag - interval;
  currentLag = (lag + (currentLag * (AVG_DECAY_FACTOR - 1))) / AVG_DECAY_FACTOR;
  lastTime = now;
}, interval);

// Don't keep process open just for this timer.
checkInterval.unref();

var toobusy = function(){
  // If current lag is < 2x the highwater mark, we don't always call it 'too busy'. E.g. with a 50ms lag
  // and a 40ms highWater (1.25x highWater), 25% of the time we will block. With 80ms lag and a 40ms highWater,
  // we will always block.
  var pctToBlock = (currentLag - highWater) / highWater;
  var rand = Math.random();
  return rand < pctToBlock;
};

toobusy.lag = function(){
  return parseInt(currentLag, 10);
};

toobusy.maxLag = function(newLag){
  if(!newLag) return highWater;

  // If an arg was passed, try to set highWater.
  if(Object.prototype.toString.call(newLag) !== "[object Number]"){
    throw "Expected numeric first argument.";
  }
  newLag = parseInt(newLag, 10);
  if(newLag < 10){
    throw "Maximum lag should be greater than 10ms.";
  }
  highWater = newLag;
  return highWater;
};

toobusy.shutdown = function(){
  clearInterval(checkInterval);
};

module.exports = toobusy;

examples/http.js
var http = require('http'),
 toobusy = require('..');

console.log("Maximum allowed event loop lag: " + toobusy.maxLag(50) + "ms");

function processRequest(res, num, startTime) {
  if (!startTime) startTime = new Date();

  if (num === undefined) {
    return process.nextTick(function() {
      processRequest(res, 0);
    });
  }
  if (num >= 5) {
    res.writeHead(200, {'Content-Type': 'text/plain'});
    res.end('I counted to ' + num + ' in ' + (new Date() - startTime) + 'ms\n');
  } else {
    // 1ms of computation
    var targetTime = (new Date() - startTime) + 1;
    while (new Date() - startTime < targetTime);
    processRequest(res, num + 1, startTime);
  }
}

http.createServer(function (req, res) {
  if (toobusy()) {
    res.writeHead(503);
    return res.end();
  }

  // we're not too busy!  let's process a request!
  processRequest(res);
}).listen(3000, '127.0.0.1', 2048);


examples/load.js
var http = require('http');

// a little load generate client that generates constant load over
// even if the server cannot keep up

var running = 0;
var twoHundred = 0;
var fiveOhThree = 0;
var yucky = 0;
var avg = 0;

// how many requests per second should we run?
const rps = (process.env['RPS'] || 40) / 40;
var curRPS = rps;
var started = 0;
const startTime = new Date();
var lastMark = startTime;

var ivalnum = 0;

setInterval(function() {
  ivalnum++;
  function startOne() {
    started++;
    running++;
    var start = new Date();
    var endOrError = false;
    function cEndOrError() {
      if (endOrError) console.log("end AND error");
      endOrError = true;
    }
    var req = http.get({
      host: '127.0.0.1',
      port: 3000,
      agent: false,
      path: '/',
      headers: {
        "connection": "close"
      }
    }, function(res) {
      if (res.statusCode === 503) {
        fiveOhThree++;
      } else {
        twoHundred++;
      }
      avg = ((new Date() - start) + avg * started) / (started + 1);
      running--;
      cEndOrError();
    }).on('error', function(e) {
      process.stderr.write(e.toString() + " - " + (new Date() - start) + "ms\n");
      avg = ((new Date() - start) + avg * started) / (started + 1);
      running--;
      yucky++;
      cEndOrError();
    });
  }

  for (var i = 0; i < curRPS ; i++) startOne();

  // report and scale up every 2s
  if (!(ivalnum % (40 * 2))) {
    var delta = (new Date() - lastMark) / 1000.0 ;
    console.log(Math.round((new Date() - startTime) / 1000.0),
                Math.round(started / delta),
                Math.round(twoHundred / delta),
                Math.round(fiveOhThree / delta),
                avg,
                Math.round(yucky / delta));
    curRPS = curRPS + .5;
    started = twoHundred = fiveOhThree = yucky = 0;
    lastMark = new Date();
  }
}, 25);
